### PR TITLE
Add Username Recovery multichannel support configs

### PR DESCRIPTION
**Pull request by Malith-19**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]
- Add separate configs for the email and sms channel in username recovery.

### Related issues
- https://github.com/wso2/product-is/issues/21180

### Important notes
- We are not removing the previous username enable config which is used to disable the connector when both email and sms channels are disabled.
- To keep the backward compatibility we are using default value of the email channel as username enable value.

### Related PRs
- https://github.com/wso2/carbon-identity-framework/pull/6040
- https://github.com/wso2/identity-apps/pull/7004
- https://github.com/wso2/identity-apps/pull/7005

### CODE REVIEW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/IdentityRecoveryConstants.java [hunk 1]
```diff
@@ -615,6 +615,8 @@ public static class ConnectorConfig {
         public static final String FORCE_ADD_PW_RECOVERY_QUESTION = "Recovery.Question.Password.Forced.Enable";
         public static final String FORCE_MIN_NO_QUESTION_ANSWERED = "Recovery.Question.MinQuestionsToAnswer";
         public static final String USERNAME_RECOVERY_ENABLE = "Recovery.Notification.Username.Enable";
+        public static final String USERNAME_RECOVERY_EMAIL_ENABLE = "Recovery.Notification.Username.Email.Enable";
```
Comment (PasinduYeshan) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/IdentityRecoveryConstants.java:618]: Is it possible to send recovery notifications simultaneously via both email and SMS channels?

**Comment by PasinduYeshan** [comment]
Update unit tests for this class.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/IdentityRecoveryConstants.java (hunk 1 above)
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/IdentityRecoveryConstants.java:618]: admins can enable both channels but when we come to the flow user should select a one channel if both were enable and if only one channel available, then notification will be send to that channel directly.

**Review by sadilchamishka — DISMISSED** [review]

**Review by sadilchamishka — APPROVED** [review]

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/IdentityRecoveryConstants.java [hunk 2]
```diff
@@ -615,6 +615,8 @@ public static class ConnectorConfig {
         public static final String FORCE_ADD_PW_RECOVERY_QUESTION = "Recovery.Question.Password.Forced.Enable";
         public static final String FORCE_MIN_NO_QUESTION_ANSWERED = "Recovery.Question.MinQuestionsToAnswer";
         public static final String USERNAME_RECOVERY_ENABLE = "Recovery.Notification.Username.Enable";
```
Comment (Malith-19) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/IdentityRecoveryConstants.java:617]: check if we need to governance apis.

**Comment by nilasini** [comment]
As I noticed for SMS based password recovery we have handled the configuration in a different way[1]. Isn't it inconsistence?

[1] https://github.com/wso2/product-is/issues/21357

**Comment by Malith-19** [comment]
> Update the logic of handling the username recovery configs.

Followed the same ondemand migration from https://github.com/wso2-extensions/identity-governance/pull/865/commits/a7a13d5e4c90fb46b46a8b8985fc701ac011126d

File: components/org.wso2.carbon.identity.governance/src/test/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                                                              boolean usernameSmsEnable) {
+
+        HashMap<String, String> expected = new HashMap<>();
+        expected.put(USERNAME_RECOVERY_ENABLE, usernameEnable ? TRUE_STRING : FALSE_STRING);
+        expected.put(USERNAME_RECOVERY_EMAIL_ENABLE, usernameEmailEnable ? TRUE_STRING : FALSE_STRING);
+        expected.put(USERNAME_RECOVERY_SMS_ENABLE, usernameSmsEnable ? TRUE_STRING : FALSE_STRING);
+
+        return expected;
+    }
+}
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.governance/src/test/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImplTest.java:256]: Add a new line.

Comment (Malith-19) [reply components/org.wso2.carbon.identity.governance/src/test/java/org/wso2/carbon/identity/governance/IdentityGovernanceServiceImplTest.java]: Fixed by https://github.com/wso2-extensions/identity-governance/pull/865/commits/cf580388de322f4d0849816cd6f0f26178e29f12

**Review by ashensw — APPROVED** [review]

**Review by ashensw — APPROVED** [review]

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/connector/RecoveryConfigImpl.java [hunk 4]
```diff
@@ -272,6 +280,10 @@ public Properties getDefaultPropertyValues(String tenantDomain) throws IdentityG
                 ConnectorConfig.RECOVERY_QUESTION_PASSWORD_RECAPTCHA_MAX_FAILED_ATTEMPTS);
         String usernameRecovery = IdentityUtil.getProperty(
                 IdentityRecoveryConstants.ConnectorConfig.USERNAME_RECOVERY_ENABLE);
+        String usernameRecoveryEmail = IdentityUtil.getProperty(
+                IdentityRecoveryConstants.ConnectorConfig.USERNAME_RECOVERY_EMAIL_ENABLE);
```
Comment (lashinijay) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/connector/RecoveryConfigImpl.java:284]: Shall we update this variable name to suggest that it's holding a boolean value?

Suggestion: isUsernameRecoveryEmailEnabled

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/connector/RecoveryConfigImpl.java [hunk 5]
```diff
@@ -272,6 +280,10 @@ public Properties getDefaultPropertyValues(String tenantDomain) throws IdentityG
...
+        String usernameRecoverySMS = IdentityUtil.getProperty(
+                IdentityRecoveryConstants.ConnectorConfig.USERNAME_RECOVERY_SMS_ENABLE);
```
Comment (lashinijay) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/connector/RecoveryConfigImpl.java:286]: Same suggestions as above

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java [hunk 6]
```diff
@@ -108,6 +109,107 @@ public static UserAccountRecoveryManager getInstance() {
...
+    public RecoveryChannelInfoDTO retrieveUsersRecoveryInformationForUsername(Map<String, String> claims,
+                                                                              String tenantDomain,
+                                                                              Map<String, String> properties)
+            throws IdentityRecoveryException {
+
+        RecoveryScenarios recoveryScenario = RecoveryScenarios.USERNAME_RECOVERY;
+        // Retrieve the user who matches the given set of claims.
+        ArrayList<org.wso2.carbon.user.core.common.User> resultedUserList = getUserListByClaims(claims, tenantDomain);
+
+        if (!resultedUserList.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java:129]: +1

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java [hunk 7]
```diff
@@ -108,6 +109,107 @@ public static UserAccountRecoveryManager getInstance() {
...
+            }
+            addRecoveryDataObject(usernameCombined.toString(), tenantDomain, recoveryFlowId, recoveryCode,
+                    recoveryScenario,
+                    notificationChannelList);
+
+            return buildUserRecoveryInformationResponseDTO(username, recoveryFlowId, recoveryCode,
+                    notificationChannelDTOS);
+
+        } else {
+            if (log.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java:206]: Fixed by https://github.com/wso2-extensions/identity-governance/pull/947

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java [hunk 8]
```diff
@@ -108,6 +109,107 @@ public static UserAccountRecoveryManager getInstance() {
...
+                } else {
+                    if (usernameCombined.length() > 0) {
+                        usernameCombined.append(",");
+                    }
+                    usernameCombined.append(username);
+                }
+            }
+            if (StringUtils.isBlank(usernameCombined.toString())) {
+                if (log.isDebugEnabled()) {
+                    log.debug("No valid user found for the given claims");
```
[the remark this answers is not in the document]
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java:193]: Fixed by https://github.com/wso2-extensions/identity-governance/pull/947

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java [hunk 9]
```diff
@@ -108,6 +109,107 @@ public static UserAccountRecoveryManager getInstance() {
...
+                            RecoverySteps.SEND_RECOVERY_INFORMATION, notificationChannelList);
+                } else {
+                    if (usernameCombined.length() > 0) {
+                        usernameCombined.append(",");
+                    }
+                    usernameCombined.append(username);
+                }
+            }
+            if (StringUtils.isBlank(usernameCombined.toString())) {
+                if (log.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java:192]: Fixed by https://github.com/wso2-extensions/identity-governance/pull/947

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java [hunk 10]
```diff
@@ -108,6 +109,107 @@ public static UserAccountRecoveryManager getInstance() {
...
+                    notificationChannels = getInternalNotificationChannelList(username, tenantDomain,
+                            recoveryScenario);
+                } else {
+                    notificationChannels = getExternalNotificationChannelList();
+                }
+
+                // Validate whether the user account is eligible for account recovery.
+                checkUserValidityForAccountRecovery(user, recoveryScenario, notificationChannels, properties);
+                // This flow will be initiated only if the user has any verified channels.
+                notificationChannelDTOS = getNotificationChannelsResponseDTOList(
```
[the remark this answers is not in the document]
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java:167]: Fixed by https://github.com/wso2-extensions/identity-governance/pull/947

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java [hunk 11]
```diff
@@ -108,6 +109,107 @@ public static UserAccountRecoveryManager getInstance() {
...
+            NotificationChannelDTO[] notificationChannelDTOS = null;
+
+            for (org.wso2.carbon.user.core.common.User resultedUser : resultedUserList) {
+                username = resultedUser.getUsername();
+                User user = Utils.buildUser(username, tenantDomain);
+
+                try {
+                    // If the account is locked or disabled, do not let the user, recover the account.
+                    checkAccountLockedStatus(user);
+
```
[the remark this answers is not in the document]
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java:147]: Fixed by https://github.com/wso2-extensions/identity-governance/pull/947

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java:129]: Fixed by https://github.com/wso2-extensions/identity-governance/pull/947

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/connector/RecoveryConfigImpl.java (hunk 5 above)
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/connector/RecoveryConfigImpl.java:286]: Fixed by https://github.com/wso2-extensions/identity-governance/pull/947

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/connector/RecoveryConfigImpl.java (hunk 4 above)
Comment (Malith-19) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/connector/RecoveryConfigImpl.java:284]: Fixed by https://github.com/wso2-extensions/identity-governance/pull/947

**Pull request merged**